Python app and dependencies

nottrobin opened the pull request:
Add a python app for serving the docs.

Fixes https://github.com/canonical-websites/docs.ubuntu.com/issues/99

QA
--


First build docs. This will take a *very* long time. Check back after 30 mins.

``` bash
documentation-builder --base-directory documents --output-path templates
```

(If you get impatient you can inspect its progress by comparing `find documents -name '*.md' | wc -l` with `find templates/en/ -name '*.html' | wc -l`)

Install dependencies

``` bash
python3 -m venv env
env/bin/pip install -r requirements.txt
```

Run the server

```
FLASK_APP=app.py env/bin/flask run
```

Now go to <http://127.0.0.1:5000> and check it looks like <https://docs.ubuntu.com/phone/>.

nottrobin committed:
Python app and dependencies

nottrobin committed:
Move documents into a documents folder

nottrobin committed:
Move app files to top level

nottrobin committed:
Ignore the right directories

nottrobin committed:
Add metadata to documents

nottrobin committed:
Host the app at /phone

nottrobin committed:
Redirecting / should be 302

nottrobin committed:
Improve finder to find any language or version

tbille commented on routing.py:
```diff
@@ -0,0 +1,223 @@
+    versions = get_versions()
+
+    if versions:
+        return version in versions
```
``` python
else:
    return False
```

tbille commented on routing.py:
```diff
@@ -0,0 +1,223 @@
+
+        return available_languages
+
+    def find_alternate_path(self, request_path, languages, versions):
```
This is a really long function, is it possible to split it in small functions?
For example a function per alternative path?

nottrobin committed:
Improvements to flask app

- Handle redirects for / and (index).html inside the app
- Allow location of redirect and version files to be configured
- Read redirects as an ordered dictionary - the order of redirects
  matters

nottrobin committed:
No longer any need for redirects in YAML

nottrobin committed:
Split up find_alternate_path into functions

nottrobin committed:
Make "is_version" always return a boolean

nottrobin commented:
@tbille I think I've addressed your concerns. Could you take another look and let me know if it's okay now?

tbille commented on routing.py:
```diff
@@ -0,0 +1,272 @@
+            template = self.templates_dir + get_file(path)
+            alt_template = self.templates_dir + get_file(alt_path)
+
+            if os.path.isfile(template):
```
I don't really like "breaking" a `for loop` with a return.
I think here there is a problem, can't you put the `if statement` outside of the loop? Otherwise this means the array `search_paths` is not constructed well

nottrobin replied:
Why don't you like breaking a for loop? I think it's pretty clear in this case. And not that it will probably matter all that much, but it's actually more efficient. Every new iteration of the loop does a new filesystem look-up, so we should really break out of the loop as soon as possible.

I don't understand your point about the array not being constructed - the array is not modified during this for loop. It's already complete before this loop starts.

tbille replied:
okay I understand we can get in efficiency by doing this. I just think it is always harder to maintain a code like this. 

(my bad for the `search_paths` I understand better what it is doing)

nottrobin replied:
Okay. I understand the concern - return statements in the middle of functions can be confusing. But in this case, the return is still at the end of the function, and as the point of the function is to search for stuff I think this is quite simple to understand.

I'd rather not have hard-and-fast rules in things like this and instead always look for examples of things that are written in overly complex ways.

In this case I think this is the simplest way to write this function.

WillMoggridge replied:
I'm a fan of returning a function at its earliest point and try to where I can.
To me, it makes the logic more explicit. If it gets to this point, it exits the function. Otherwise you need to scan ahead and check what happens.

I think returning early actually makes the code more maintainable and easier to read. If this is ever not the case then it could be split up more, or it can be flexible and return later. As Robin put it, I like it not being a hard and fast rule.

tbille commented on routing.py:
```diff
@@ -0,0 +1,272 @@
+
+                if lines:
+                    for url_match, target_url in lines.items():
+                        # Make sure all matches start with slash
```
Useless comment

nottrobin replied:
fixed

tbille commented on routing.py:
```diff
@@ -0,0 +1,272 @@
+                parts = {}
+                for name, value in result.groupdict().items():
+                    parts[name] = value or ''
+                return target.format(**parts)
```
same here is possible to put the return out of the for loop?

nottrobin replied:
The same applies here. The redirects file can get thousands of lines long and regex matches are among some of the most expensive operations in Python, so I want to break the loop ASAP.

tbille commented:
And one last thing. Since here it is all in the same PR. It could be cleaner to squash some commits (for example the last `Remove "useless comment"`) This should be on a git workflow, commits help a future dev to understand the choices made on a project, I don't think (s)he will care about this kind of commit :smile:

nottrobin commented:
@tbille If you feel strongly about the commit point we should discuss it with the team in dev catch-up. Personally, I don't like the idea of squashing this whole PR into one commit - I think it's more useful that "Improve finder to find any language or version" remain its own commit than that 'Remove "useless comment"' is not its own commit - I think people reading history can easily skip over useless commits, but merging separate pieces of work into one commit is actually losing information.

The alternative is to completely rewrite my commits into the relevant pieces of work just before I merge, which I sometimes do, but is time-consuming so I don't think we should force everyone to do that every time. 

For now I'm not going to squash or rewrite my commits I'm afraid.

nottrobin commented:
After some discussion with @tbille, I misunderstood what he was asking for. The "useless comment" commit has been removed now. :grinning:

nottrobin commented:
I tried playing with rewritetarget. The main problem with it was redirects: If you have redirects inside your app that redirect e.g. `/index.html` -> `/`, then with rewritetarget it will redirect `/phone/index.html` to `/`, which is not right.

I couldn't find an easy way around that. So it seemed neatest to just make the app aware of its subpath

@WillMoggridge If you have a better solution, please let me know

WillMoggridge commented:
Oops... I think I just accidentally deleted my review which was:
> Looks good.
>
> It is missing a 404 page though, I think we want one?
>
> Another note is that I see the /phone path is hardcoded. I'm happy with that but in case you haven't seen it we also have the option to serve it on the root and Kubernetes can handle the logic. This could allow us to keep the app simpler and move between subdomain or subfolder with no code changes.
The annotation for it is ingress.kubernetes.io/rewrite-target: /

WillMoggridge commented:
Yeah, that is a good point I didn't think about.
I don't know a good solution to that for now, so this sounds like a good plan.

WillMoggridge reviewed (approved):
Looks good then.

I have created an issue for the 404 page at https://github.com/canonical-docs/phone-docs/issues/14
